src/s_stuff.h

```diff
@@ -185,6 +192,16 @@ EXTERN void sys_closesocket(int fd);
 typedef void (*t_fdpollfn)(void *ptr, int fd);
 EXTERN void sys_addpollfn(int fd, t_fdpollfn fn, void *ptr);
 EXTERN void sys_rmpollfn(int fd);
+#ifdef THREADED_IO
+#define t_rbskt struct _rbskt
+typedef void (*t_fdsendrmfn)(void *ptr);
+EXTERN void sys_addpollrb(int fd, int preserve_boundaries);
+EXTERN t_rbskt* sys_getpollrb(int fd);
```

Comment:
I was wondering whether `sys_addpollrb` could return a `t_rbskt *`. As an external author, I would rather store the `t_rbskt *` once in the beginning and not call `sys_getpollrb` everytime in the poll function. Actually, `sys_getpollrb` might not be necessary at all...

Reply:
That could work, however how do you guarantee that he lifetime of the `t_rbskt*` is respected? It gets destroyed when `sys_rmpollfn()` is called, the object will have to know that they can no longer use it after that.

In a previous version of the code I had the `t_rbskt` owned by the object and not by `s_inter`, but that was causing issues at cleanup, so I opted for this version in the end.

Reply:
> It gets destroyed when sys_rmpollfn() is called

Isn't `sys_rmpollfn()` always called by the client (in the `t_fdsendrmfn` or when it encounters an error in `recv`)? I don't see how Pd itself would remove the fd from the list. I think as long as the client doesn't call, `sys_rmpollfn`, the ringbuffer should be valid.

Reply:
Of course, the ring buffer should still be owned by Pd and the client only gets an opaque pointer.

Reply:
Actually this is not straightforward to do: it's easy to have `sys_addpollrb()` return a  `t_rbskt *`, and it's straightforward to store that inside `t_netsend` so that it can be used by `netsend_readbin()`. However, when the `pollfn` that is registered is `socketreceiver_read`, there is no easy way to store that value inside `socketreceiver_t`, because the latter is opaque for the code in `x_net.c`

Two possible workarounds:
- `socketreceiver_read()` could call `sys_getpollrb()` upon first invocation and then cache it internally
- a new public function e.g.:`socketereceiver_setrbskt()` that can be called from within `x_net.c` with the value obtained from calling `sys_addpollrbskt()`

thoughts?

Reply:
I see. I'll have a closer look.

Reply:
@umlaeute Any thoughts on this? What would you prefer if you were to switch `iemnet` to the API? ;-)

Reply:
i wouldn't ever use a conditional API.
please get rid of `#ifdef THREADED_IO` in all header files, and make sure that the new functions work, regardless of whether Pd was compiled with threaded io or without.

i haven't had a closer look at the rest, but it seems you are going along nicely ;-)

Reply:
`sys_addpollrb(int fd, int preserve_boundaries)` could signal success or failure, either by returning an `int` (true/false) or by returning `NULL` (in case you pick up my idea of returning a `t_rbskt*`). Then the external would know what it has to do in the poll function (`rbskt_recv` vs `recv`)

Reply:
Good idea. I never meant to leave it conditional as it is now. This was for testing logic and functionality and provide clear separation of concerns that would make it easier to review the design.

Reply:
bd900983:

All threaded behaviour becomes a runtime option. Currently this is hardcoded as enabled, but it could be set optionally via flags (startup flags for GUI, additional arguments for `[netsend]`/`[netreceive]`).

Reply:
This looks good now!